File: Trie.java

// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No issues


// Your code here along with comments explaining your approach

class Trie {
    class TrieNode{
        boolean isEnd;
        TrieNode[] children;
        public TrieNode(){
            this.isEnd = false;
            this.children = new TrieNode[26];
        }
    }

    public TrieNode root;

    public Trie() {
        root = new TrieNode();
    }
    // Time Complexity : O(l)
    // Space Complexity : O(l)
    public void insert(String word) {
        TrieNode curr = root;
        for(int i = 0 ; i < word.length(); i++){
            char c = word.charAt(i);
            if(curr.children[c - 'a'] == null){
                curr.children[c - 'a'] = new TrieNode();
            }
            curr = curr.children[c - 'a']; // move curr pointer to next child
        }

        curr.isEnd = true;
    }
    // Time Complexity : O(l)
    // Space Complexity : O(1)
    public boolean search(String word) {
        TrieNode curr = root;
        for(int i = 0 ; i < word.length(); i++){
            char c = word.charAt(i);
            if(curr.children[c - 'a'] == null){ // we don't have next child, word is not present
                return false;
            }
            curr = curr.children[c - 'a']; // move curr pointer to next child
        }

        return curr.isEnd;
    }
    // Time Complexity : O(l)
    // Space Complexity : O(1)
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for(int i = 0 ; i < prefix.length(); i++){
            char c = prefix.charAt(i);
            if(curr.children[c - 'a'] == null){ // we don't have next child, prefix is not present
                return false;
            }
            curr = curr.children[c - 'a']; // move curr pointer to next child
        }

        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: trie.cpp
// Time complexity- insert(O(word-length)) search(O(n)) startswith-O(n)
// space complexity O(n*k*n);
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no

// create a class with array of size 26 to store each character of a word
// have a variable to help keep track of when the word ends
// the array helps to enable a constant look up for each alphabet


#include<algorithm>
#include<iostream>
#include <sstream>
#include<vector>
#include<stack>

using namespace std;
class Trie {
public:
    
    /** Initialize your data structure here. */
    Trie() {
        
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        Trie* node =this;
        for(auto ch : word){
            ch -='a';
            if(node->child[ch]== nullptr){
                node->child[ch] = new Trie();
            }
            node = node->child[ch];
        }
        node->isEnd=true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        Trie* node = this;
        for(auto ch : word){
            ch -= 'a';
            if(node->child[ch]== nullptr){
                return false;
            }
            node = node->child[ch];
        }
        return node->isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        Trie*node =this;
        for(auto ch : prefix){
            ch-='a';
            if(node->child[ch]== nullptr){
                return false;
            }
            node = node->child[ch];
        }
        return true;
    }
private:
    Trie* child[26]={};       
    bool isEnd = false;
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */